File: interview-app-objc/interview-app-objc/CoffeeShopFinderService.m
//
//  CoffeeShopFinderService.m
//  interview-app-objc
//
//  Created by Tony Nuzzi on 5/28/15.
//  Copyright (c) 2015 Tony Nuzzi. All rights reserved.
//

#import "CoffeeShopFinderService.h"

@implementation CoffeeShopFinderService

- (instancetype)init
{
    self = [super init];
    if (self) {
        
        // Create an empty array so consuming classes don't have to nil check
        [self setCoffeeshopsInMyArea:@[]];
        
        // Create a weak reference to the delegates added, as to not create a memory leak
        [self set_delegates:[NSHashTable weakObjectsHashTable]];
        
        // Initialize Yelp API service
        [self setYelpService:[[YelpAPIService alloc] init]];
        [self.yelpService setDelegate:self];
        
        // Initialize geocoder for reverse lookup
        [self setGeocoder:[[CLGeocoder alloc] init]];
        
        // Start initialization of CoreLocation Manager
        [self setLocationManager:[[CLLocationManager alloc] init]];
        [self.locationManager setDelegate:self];
        
        // Set desired accuracy to "nearest ten meters".  Does not need to best accuracy
        [self.locationManager setDesiredAccuracy:kCLLocationAccuracyNearestTenMeters];
        
        // Check to make sure location services are enabled on user's device
        if([CLLocationManager locationServicesEnabled]) {
            
            // If authorization status is not determined ask user permission
            if([CLLocationManager authorizationStatus] == kCLAuthorizationStatusNotDetermined) {
                [self.locationManager requestWhenInUseAuthorization];
                
                // If user denied/rejected or is the wrong authorization; display alert with button to settings URL
            } else if([CLLocationManager authorizationStatus] == kCLAuthorizationStatusDenied ||
                      [CLLocationManager authorizationStatus] == kCLAuthorizationStatusRestricted ||
                      [CLLocationManager authorizationStatus] == kCLAuthorizationStatusAuthorizedAlways) {
                
                [self showSettingAlertDialog];
            }
        }
    }
    return self;
}

-(void) findCoffeeshopsInMyArea {
    if([CLLocationManager authorizationStatus] == kCLAuthorizationStatusAuthorizedWhenInUse) {
        [self.locationManager startUpdatingLocation];
    } else if([CLLocationManager authorizationStatus] == kCLAuthorizationStatusDenied ||
              [CLLocationManager authorizationStatus] == kCLAuthorizationStatusRestricted ||
              [CLLocationManager authorizationStatus] == kCLAuthorizationStatusAuthorizedAlways) {
        [self showSettingAlertDialog];
    }
}

- (void) stop {
    [self.locationManager stopUpdatingLocation];
}

- (void) resume {
    [self.locationManager startUpdatingLocation];
}


- (void) addDelegate: (id<CoffeeShopFinderServiceDelegate>) delegate
{
    [self._delegates addObject: delegate];
}

// Calling this method is optional.
// The hash table will automatically remove the delegate when it gets released
- (void) removeDelegate: (id<CoffeeShopFinderServiceDelegate>) delegate
{
    [self._delegates removeObject: delegate];
}

#pragma mark - Private

- (void) showSettingAlertDialog {
    // Initialize alert controller
    UIAlertController* alertController = [UIAlertController alertControllerWithTitle:@"Background Location Access Disabled" message:@"In order to find coffee shops around you, please open this app's settings and set location access to 'Always'." preferredStyle:UIAlertControllerStyleAlert];
    
    // Cancel Action button
    UIAlertAction* cancelAction = [UIAlertAction actionWithTitle:@"Cancel" style:UIAlertActionStyleCancel handler:nil];
    [alertController addAction:cancelAction];
    
    // Open Settings action button
    UIAlertAction* openAction = [UIAlertAction actionWithTitle:@"Open Settings" style:UIAlertActionStyleDefault handler:^(UIAlertAction *action){
        NSURL* settingUrl = [NSURL URLWithString:UIApplicationOpenSettingsURLString];
        if (settingUrl != nil) {
            [[UIApplication sharedApplication] openURL:settingUrl];
        }
    }];
    [alertController addAction:openAction];
    
    // Show Alert controller
    // NOTE: Since this code is not run in UIViewController use the root view controller to present the alert
    [[[[[UIApplication sharedApplication] delegate] window] rootViewController] presentViewController:alertController animated:YES completion:nil];
}

#pragma mark - CLLocationManagerDelegate

// Using deletage to start listening to updates for the location
- (void) locationManager:(CLLocationManager *)manager didChangeAuthorizationStatus:(CLAuthorizationStatus)status {
    if (status == kCLAuthorizationStatusAuthorizedWhenInUse) {
        [self.locationManager startUpdatingLocation];
    }
}

- (void) locationManager:(CLLocationManager *)manager didUpdateLocations:(NSArray *)locations {
    // Fail fast if there are no locations
    if([locations count] == 0) {
        return;
    }
    
    // Get the first location (User's current location)
    CLLocation* location = [locations objectAtIndex:0];
    
    [self setMyLatitude: location.coordinate.latitude];
    [self setMyLongitude: location.coordinate.longitude];
    
    [self.geocoder reverseGeocodeLocation:location completionHandler:^(NSArray* placemarks, NSError* error){
        // Fail fast if we have errors or can't reverse lookup location
        if(error != nil) {
            NSDictionary *userInfo = [error userInfo];
            NSString *errorString = [[userInfo objectForKey:NSUnderlyingErrorKey] localizedDescription];
            NSLog(@"%@", errorString);
            
            [[[UIAlertView alloc] initWithTitle:@"Error"
                                        message:@"Failed to lookup location address"
                                       delegate:nil
                              cancelButtonTitle:@"OK"
                              otherButtonTitles:nil] show];
            return;
        }
        
        if([placemarks count] == 0) {
            return;
        }
        
        NSDictionary* dict = [[placemarks objectAtIndex:0] addressDictionary];
        
        NSString* searchLocation = [NSString stringWithFormat:@"%@ %@ %@",
                                    [dict objectForKey:(NSString *) kABPersonAddressStreetKey],
                                    [dict objectForKey:(NSString *) kABPersonAddressCityKey],
                                    [dict objectForKey:(NSString *) kABPersonAddressZIPKey]];
        
        [self.yelpService findNearByCoffeshopsByLocation:searchLocation aLatitude:location.coordinate.latitude andLongitude:location.coordinate.longitude];
    }];
    
    // Just get the first location update then stop
    [self.locationManager stopUpdatingLocation];
}

#pragma mark - YelpAPIServiceDelegate

- (void)loadResultWithDataArray:(NSArray *)resultArray {
    
    // Cache results
    self.coffeeshopsInMyArea = [NSArray arrayWithArray:resultArray];
    
    // Call all registered delegates
    for(id<CoffeeShopFinderServiceDelegate> delegate in self._delegates){
        [delegate coffeeshopsInMyArea:self.coffeeshopsInMyArea myLatitude:self.myLatitude andLongitude:self.myLongitude];
    }
}
@end

File: interview-app-swift/interview-app-swift/CoffeeShopFinderService.swift
//
//  CoffeeShopFinderService.swift
//  interview-app-swift
//
//  Created by Tony Nuzzi on 6/2/15.
//  Copyright (c) 2015 Tony Nuzzi. All rights reserved.
//

import Foundation
import CoreLocation
import UIKit
import AddressBook

protocol CoffeeShopFinderServiceDelegate {
    func coffeeShopsInMyArea(coffeeShops: [CoffeeshopModel], myLatitiude latitude: CLLocationDegrees,
        andLongitude longitude: CLLocationDegrees)
    func noCoffeeshopsFound(atLatitude: CLLocationDegrees, andLongitude: CLLocationDegrees)
}

class CoffeeShopFinderService: NSObject, CLLocationManagerDelegate, YelpAPIServiceDelegate {
    // Array to hold the CoffeeshopModels
    var coffeeShopsInMyArea: [CoffeeshopModel] = []
    
    // LocationManager and Geocoder for getting location and reverse geolocation lookup
    var locationManager: CLLocationManager?
    var geocoder: CLGeocoder?
    
    // Help service helper
    var yelpAPIService: YelpAPIService?
    
    // User's current location
    var myLatitude: CLLocationDegrees?
    var myLongitude: CLLocationDegrees?
    
    // Array of registered delegates
    private var _delegates: [CoffeeShopFinderServiceDelegate] = []
    
    override init() {
        super.init()
        
        // Initialize Yelp service class and assign delegate
        self.yelpAPIService = YelpAPIService()
        self.yelpAPIService?.delegate = self;
        
        // Initalize Geocoder
        self.geocoder = CLGeocoder()
        
        // Initialize LocationManager and assign delegate
        self.locationManager = CLLocationManager()
        self.locationManager?.delegate = self
        
        // Measure geolocation accuracy to the ten meters (wanting quickness rather than accuracy)
        self.locationManager?.desiredAccuracy = kCLLocationAccuracyNearestTenMeters
        
        // Check with location manager to see if user has authorized app to get location information
        // If not show alert
        if CLLocationManager.locationServicesEnabled() {
            if CLLocationManager.authorizationStatus() == .NotDetermined {
                self.locationManager!.requestWhenInUseAuthorization()
            } else if CLLocationManager.authorizationStatus() == .Denied ||
                      CLLocationManager.authorizationStatus() == .Restricted ||
                      CLLocationManager.authorizationStatus() == .AuthorizedAlways {
                self.showSettingAlertDialog()
            }
        }
    }
    
    func findCoffeeshopsInMyArea() {
        if CLLocationManager.authorizationStatus() == .AuthorizedWhenInUse {
            self.locationManager?.startUpdatingLocation()
        } else if CLLocationManager.authorizationStatus() == .Denied ||
                  CLLocationManager.authorizationStatus() == .Restricted ||
                  CLLocationManager.authorizationStatus() == .AuthorizedAlways {
            self.showSettingAlertDialog()
        }
    }
    
    // Used in app delegate when app is backgrounded
    func stop() {
        self.locationManager?.stopUpdatingLocation()
    }
    
    // Used when app is placed in the foreground
    func resume() {
        self.locationManager?.startUpdatingLocation()
    }
    
    func addDelegate(delegate: CoffeeShopFinderServiceDelegate) {
        self._delegates.append(delegate)
    }
    
    // Helper method to show alert if user has not authorized app
    private func showSettingAlertDialog() {
        // Initialize alert controller
        var alertController = UIAlertController(title: "Background Location Access Disabled", message: "In order to find coffee shops around you, please open this app's settings and set location access to 'While Using the App'.", preferredStyle: .Alert)
        
        // Cancel Action button and add to controller
        var cancelAction = UIAlertAction(title: "Cancel", style: .Cancel, handler: nil)
        alertController.addAction(cancelAction)
        
        // Open Action button and add to controller
        var openAction = UIAlertAction(title: "Open Settings", style: .Default, handler: {
            (action: UIAlertAction!) in
            if let settingUrl = NSURL(string: UIApplicationOpenSettingsURLString) {
                UIApplication.sharedApplication().openURL(settingUrl)
            }
        })
        alertController.addAction(openAction)
        
        // Show Alert controller
        // NOTE: Since this code is not run in UIViewController use the root view controller to present the alert
        UIApplication.sharedApplication().delegate!.window!!.rootViewController!.presentViewController(alertController, animated: true, completion: nil)
    }
    
    // Mark: - CLLocationManagerDelegate
    
    func locationManager(manager: CLLocationManager!, didChangeAuthorizationStatus status: CLAuthorizationStatus) {
        if status == .AuthorizedWhenInUse {
            manager.startUpdatingLocation()
        }
    }
    
    func locationManager(manager: CLLocationManager!, didUpdateLocations locations: [AnyObject]!) {
        // Fail fast if not location are returned
        if locations.count == 0 {
            self.notifyDelegatesNoCoffeeshopsFound()
            return
        }
        
        // Location are placed in array chronologically, so last should be the latest
        var location = locations.last as! CLLocation
        
        // Setting user's location
        self.myLatitude = location.coordinate.latitude
        self.myLongitude = location.coordinate.longitude
        
        // Doing a reverse lookup to get address information
        self.geocoder!.reverseGeocodeLocation(location, completionHandler: {
            (placemarks: [AnyObject]!, error: NSError!) in
            
            // Fail fast if an error occured
            // To the user it will look like no coffeeshops are available
            if(error != nil) {
                self.notifyDelegatesNoCoffeeshopsFound()
                return
            }
            
            // If address can't not be determine notify user no coffee shops are available
            if placemarks.count == 0 {
                self.notifyDelegatesNoCoffeeshopsFound()
                return
            }
            
            // Cast array to array of CLPlacemark
            let placemarkArray = placemarks as! [CLPlacemark]
            
            // Get the address of the first location using the address dictionary
            let addressDict = placemarkArray[0].addressDictionary
            
            
            // Retrieve address information from address dictionary and concat together
            var searchLocation = "\(addressDict[kABPersonAddressStreetKey]) \(addressDict[kABPersonAddressCityKey]) \(addressDict[kABPersonAddressZIPKey])"
            
            // Call yelp service to lookup coffees shop at user's current location
            self.yelpAPIService?.findNearByCoffeshopsByLocation(searchLocation, aLatitude: self.myLatitude!, andLongitude: self.myLongitude!)
        })
        
        // Only get the first location then stop listening
        self.locationManager?.stopUpdatingLocation()
    }
    
    // Mark: - YelpAPIServiceDelegate
    func loadResultWithDataArray(resultArray: [CoffeeshopModel]) {
        // If no coffeeshops are found in user's location let them know
        if resultArray.count == 0 {
            self.coffeeShopsInMyArea = [];
            self.notifyDelegatesNoCoffeeshopsFound()
            
            return
        }
        
        // Array of coffee shop models
        self.coffeeShopsInMyArea = resultArray
        
        // Call all registered delegates
        for delegate: CoffeeShopFinderServiceDelegate in self._delegates {
            delegate.coffeeShopsInMyArea(self.coffeeShopsInMyArea, myLatitiude: self.myLatitude!, andLongitude: self.myLongitude!)
        }
    }
    
    // Mark: - Private
    private func notifyDelegatesNoCoffeeshopsFound() {
        
        // Make sure cached version is set to empty
        self.coffeeShopsInMyArea =  [];
        
        // Initialize alert controller
        var alertController = UIAlertController(title: "Sorry", message: "No coffee shops were found in your area.", preferredStyle: .Alert)
        
        // Ok Action button and add to controller
        var okAction = UIAlertAction(title: "Ok", style: .Default, handler: nil)
        alertController.addAction(okAction)
        
        // Show Alert controller
        UIApplication.sharedApplication().delegate!.window!!.rootViewController!.presentViewController(alertController, animated: true, completion: nil)
        
        // Call all registered delegates
        for delegate: CoffeeShopFinderServiceDelegate in self._delegates {
            delegate.noCoffeeshopsFound(self.myLatitude!, andLongitude: self.myLongitude!)
        }
    }
}